[CLI] Add minimal Ledger key rotation support

alnoki opened the pull request:
@banool @davidiw @gedigi @gregnazario @hardsetting

# Background

This PR adds CLI authentication key rotation support for Ledger-based accounts.
This enhances OpSec/DevOps in several situations:

1. Allow developers to convert legacy hot wallets (from before Ledger support
   was available) to hardware wallets, for example hot wallets that were used
   to deploy dApp Move packages.
1. Enable publication from a Ledger hardware wallet when the package size is
   above the Ledger signing limit, by temporarily rotating to a hot key then
   rotating back.
1. Enable vanity addresses secured by a hardware wallet, useful for account
   labelling in organizations that have many segmented accounts for different
   purposes.

# Contribution history

I originally proposed a large set of changes associated with authentication key
rotation for accounts secured by a Ledger wallet in #11151, which paired with a
docs PR at https://github.com/aptos-labs/developer-docs/pull/367.

However, due to the difficulty of merging large sets of changes as an external
contributor, I'm splitting off into this PR the minimum amount of functionality
required to support Ledger key rotation in the Aptos CLI.

# Verifying the changes in this PR

1. Install the Aptos CLI [from source] using the changes in this PR.

1. Connect and unlock your Ledger.

1. Check what version of the Aptos app you have: `Aptos > About > Version`.

1. If you do not have version `0.6.9` or higher, update it using Ledger Live.

1. Enable blind signing: `Aptos > Settings > Enable Blind Signing`.

1. Start the interactive profile creator:

   ```sh
   aptos init --profile devnet-hot-wallet-before-rotation
   ```

1. Select `devnet`, then click `return` to generate a new private key.

1. Rotate the authentication key of the hot wallet so that it is secured by
   your Ledger:

   ```sh
   aptos account rotate-key \
       --new-derivation-index 1000 \
       --profile devnet-hot-wallet-before-rotation \
       --save-to-profile devnet-hot-wallet-rotated-to-ledger
   ```

1. Follow the instructions to sign the rotation proof challenge signature on
   your Ledger, then approve the transaction in the CLI.

1. Generate a new private key file:

   ```sh
   aptos key generate --output-file private-key-to-rotate-to
   ```

1. Rotate the Ledger wallet to the new private key:

   ```sh
   aptos account rotate-key \
       --new-private-key-file private-key-to-rotate-to \
       --profile devnet-hot-wallet-rotated-to-ledger \
       --save-to-profile devnet-ledger-wallet-rotated-to-private-key
   ```

1. Follow the instructions to sign the rotation proof challenge signature on
   your Ledger, then approve the transaction in the CLI (you'll need to sign
   with your Ledger a second time to approve the transaction).

1. Compare the three profiles:

   ```sh
   aptos config show-profiles --profile devnet-hot-wallet-before-rotation
   aptos config show-profiles --profile devnet-hot-wallet-rotated-to-ledger
   aptos config show-profiles --profile devnet-ledger-wallet-rotated-to-private-key
   ```

# Verify normal rotation

Per [review comment] from @banool, you can verify normal rotation as per below.

1. Generate another private key:

   ```sh
   aptos key generate --output-file yet-another-private-key
   ```

1. Rotate from the final hot wallet profile above to another private key:

   ```sh
   aptos account rotate-key \
       --new-private-key-file yet-another-private-key \
       --profile devnet-ledger-wallet-rotated-to-private-key \
       --save-to-profile devnet-hot-wallet-to-hot-wallet
   ```

1. Compare the two profiles:

   ```sh
   aptos config show-profiles --profile devnet-ledger-wallet-rotated-to-private-key
   aptos config show-profiles --profile devnet-hot-wallet-to-hot-wallet
   ```

## Housekeeping

The following was run from repository root to ensure CI passes:

```sh
pre-commit run --all-files
scripts/rust_lint.sh
cargo test --package aptos
```

CLI e2e tests per README in `/crates/aptos/e2e`:

```sh
DOCKER_DEFAULT_PLATFORM=linux/amd64 poetry run python main.py \
    --base-network mainnet \
    --debug \
    --test-cli-path ~/your/local/path/to/aptos-core/target/debug/aptos
```

[from source]: https://aptos.dev/en/build/cli/install-cli/install-cli-specific-version
[review comment]: https://github.com/aptos-labs/aptos-core/pull/14084#pullrequestreview-2196459492

banool commented on crates/aptos/src/account/key_rotation.rs:
```diff
@@ -38,31 +36,57 @@ use std::{collections::BTreeMap, path::PathBuf};
+    /// BIP44 account index of hardware wallet account, e.g. `0`
     ///
-    /// This skips the interactive profile saving after rotating the authentication key
+    /// Given index `n` maps to BIP44 derivation path `m/44'/637'/n'/0'/0`
```
Neat, very helpful comments.

banool commented on crates/aptos/src/account/key_rotation.rs:
```diff
@@ -38,31 +36,57 @@ use std::{collections::BTreeMap, path::PathBuf};
+pub(crate) struct NewProfileOptions {
+    /// Only specify if you do not want to save a new profile
     #[clap(long)]
     pub(crate) skip_saving_profile: bool,
```
I know this feature already existed but when would you want to do this? cc @gregnazario

alnoki replied:
> I know this feature already existed 

Right, I don't want to break compatibility so I'm just trying to make sure the grouping semantics are workable (though I still think it could be useful, see elow)

> when would you want to do this

1. You just want an example of a transaction
1. You are running an e2e test on a persisted runner and don't want to clutter the config
1. You are rotating back to a ledger after temporarily rotating to a hot wallet in order to publish a transaction, and you don't want to clutter up your config because elsewhere you have a database of which derivation path should map to which wallet

banool commented on crates/aptos/src/account/key_rotation.rs:
```diff
@@ -94,56 +118,131 @@ impl CliCommand<RotateSummary> for RotateKey {
-        let rotation_proof_signed_by_new_private_key =
-            new_private_key.sign_arbitrary_message(&rotation_msg);
+        // Sign the struct using both the current private key and the new private key.
+        let rotation_proof_signed_by_current_private_key = if current_derivation_path.is_some() {
+            eprintln!("Approve rotation proof challenge signature on your Ledger device");
+            aptos_ledger::sign_message(
+                current_derivation_path.clone().unwrap().as_str(),
```
Nit: `if let Some` (or personally I prefer just good old `match`) to avoid the `is_some` -> `.unwrap()` pattern.

alnoki replied:
Addressed in 5b9dc70

banool reviewed (approved):
Thanks for the PR and apologies again for the delay on reviewing your other PR. I just ran through your guide on how to verify the changes in the PR and everything worked as described. I even did steps 8 through 11 again with the new profiles to double check.

After "Approve rotation proof challenge signature on your Ledger device" it might be nice if it printed out "Success, will now rotate key" or something, it felt a little awkward otherwise why I have to do two things on the device.

Just to confirm, have you verified that this doesn't cause issues with rotating normally, without a ledger?

Some of my comments are not explicitly related to this PR, I'm sorta new to this code, so trying to catch up.

As far as I can tell this seems correct, all the necessary checks in place, but someone else should check my work.

alnoki commented on crates/aptos/src/account/key_rotation.rs:
```diff
@@ -152,104 +251,65 @@ impl CliCommand<RotateSummary> for RotateKey {
-
         if let Some(txn_success) = txn_summary.success {
             if !txn_success {
+                eprintln!("{}", txn_string);
```
I'm afraid I'm not sure what you mean by this question:

> We're not hiding important information from the user now?

However, the rationale for including the print output with the txn info here is so that the user has immediate feedback on if the rotation has succeeded or not onchain, because there could theoretically be an error later with updating the config and if this were the case I figured the user should at least know the txn went through or not

banool replied:
My point is it used to be printed unconditionally, now it's not printed if the txn succeeded.

alnoki replied:
I understand your concern now, but the txn info is *still* printed in the case of success:

https://github.com/aptos-labs/aptos-core/blob/5b9dc700eea51361fc5619c74899ffffd27d39a7/crates/aptos/src/account/key_rotation.rs#L327-L332

alnoki commented on crates/aptos/src/account/key_rotation.rs:
```diff
@@ -94,56 +118,131 @@ impl CliCommand<RotateSummary> for RotateKey {
 
+        // Submit transaction.
+        if current_derivation_path.is_some() {
+            eprintln!("Approve transaction on your Ledger device");
```
Addressed in 5b9dc70

alnoki commented:
@banool per https://github.com/aptos-labs/aptos-core/pull/14084#pullrequestreview-2196459492

> After "Approve rotation proof challenge signature on your Ledger device" it might be nice if it printed out "Success, will now rotate key" or something, it felt a little awkward otherwise why I have to do two things on the device.

I've updated the printouts to be more informative per associated review comments

> Just to confirm, have you verified that this doesn't cause issues with rotating normally, without a ledger?

I've added a section `Verify normal rotation` to the PR description

banool commented:
Great, let's get one more stamp and then we're good to go (once we can get the CI to run happily).
